<html><head><title>niplav</title>
<link href="./favicon.png" rel="shortcut icon" type="image/png"/>
<link href="main.css" rel="stylesheet" type="text/css"/>
<meta content="text/html; charset=utf-8" http-equiv="Content-Type"/>
<!DOCTYPE HTML>

<style type="text/css">
code.has-jax {font: inherit; font-size: 100%; background: inherit; border: inherit;}
</style>
<script async="" src="./mathjax/latest.js?config=TeX-MML-AM_CHTML" type="text/javascript">
</script>
<script type="text/x-mathjax-config">
	MathJax.Hub.Config({
	extensions: ["tex2jax.js"],
	jax: ["input/TeX", "output/HTML-CSS"],
	tex2jax: {
		inlineMath: [ ['$','$'], ["\\(","\\)"] ],
		displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
		processEscapes: true,
		skipTags: ['script', 'noscript', 'style', 'textarea', 'pre']
	},
	"HTML-CSS": { availableFonts: ["TeX"] }
	});
</script>
<script>
document.addEventListener('DOMContentLoaded', function () {
	// Change the title to the h1 header
	var title = document.querySelector('h1')
	if(title) {
		var title_elem = document.querySelector('title')
		title_elem.textContent=title.textContent + " ‚Äì niplav"
	}
});
</script>
</head><body><h2 id="home"><a href="./index.html">home</a></h2>
<p><em>author: niplav, created: 2022-01-25, modified: 2022-02-23, language: english, status: in progress, importance: 2, confidence: likely</em></p>
<blockquote>
<p><strong>Some solutions to exercises in the book ‚ÄúPopulation Genetics‚Äù by
<a href="https://en.wikipedia.org/wiki/John_H._Gillespie">John H. Gillespie</a>. I
did not simply copy out the solutions at the end of each chapter, and
sometimes didn‚Äôt even check the solutions against my own. Therefore,
these might be faulty.</strong></p>
</blockquote>
<h1 id="Solutions_to_Population_Genetics"><a class="hanchor" href="#Solutions_to_Population_Genetics">Solutions to ‚ÄúPopulation Genetics‚Äù</a></h1>
<h2 id="Chapter_1"><a class="hanchor" href="#Chapter_1">Chapter 1</a></h2>
<h3 id="Problem_11"><a class="hanchor" href="#Problem_11">Problem 1.1</a></h3>
<p>For three it would be <code>$A_1A_1A_1, A_1A_1A_2, A_1A_2A_2, A_1A_2A_3, A_2A_2A_2, A_2A_2A_3, A_3A_3A_3$</code>,
with 7 different genotypes.</p>
<p>For four it would be 32 different genotypes:</p>
<p><code>$A_1 A_1 A_1 A_1, A_1 A_1 A_1 A_2, A_1 A_1 A_1 A_3, A_1 A_1 A_1 A_4, A_1 A_1 A_2 A_2$</code>,
<code>$A_1 A_1 A_2 A_3, A_1 A_1 A_2 A_4, A_1 A_1 A_3 A_3, A_1 A_1 A_3 A_4, A_1 A_1 A_4 A_4$</code>,
<code>$A_1 A_2 A_2 A_2, A_1 A_2 A_2 A_3, A_1 A_2 A_2 A_4, A_1 A_2 A_3 A_4, A_1 A_2 A_4 A_4$</code>,
<code>$A_1 A_3 A_4 A_4, A_1 A_4 A_4 A_4, A_2 A_2 A_2 A_2, A_2 A_2 A_2 A_3, A_2 A_2 A_2 A_4$</code>,
<code>$A_2 A_2 A_3 A_3, A_2 A_2 A_3 A_4, A_2 A_2 A_4 A_4, A_2 A_3 A_3 A_3, A_2 A_3 A_3 A_4$</code>,
<code>$A_2 A_3 A_4 A_4, A_2 A_4 A_4 A_4, A_3 A_3 A_3 A_3, A_3 A_3 A_3 A_4, A_3 A_3 A_4 A_4$</code>,
<code>$A_3 A_4 A_4 A_4, A_4 A_4 A_4 A_4$</code>.</p>
<p>So apparently, per solution, this is wrong, because I assumed that <code>$n$</code>
different alleles resulted in a <code>$n$</code>ploid organism, which isn‚Äôt the case.</p>
<p>For three alleles it would be <code>$A_1 A_1, A_1 A_2, A_1 A_3, A_2 A_2, A_2 A_3, A_3 A_3$</code>,
which is 6 different genotypes.</p>
<p>For four different alleles it would be 10 different genotypes:</p>
<p><code>$A_1 A_1,A_1 A_2,A_1 A_3,A_1 A_4,A_2 A_2,A_2 A_3,A_2 A_4,A_3 A_3,A_3 A_4,A_4 A_4$</code></p>
<p>The general formula is <code>$\frac{n (n+1)}{2}$</code> different genotypes,
because it‚Äôs just the upper half of the square again.</p>
<h3 id="Problem_12"><a class="hanchor" href="#Problem_12">Problem 1.2</a></h3>
<ul>
<li>For I: <code>$0.0151+\frac{1}{2}0.0452+\frac{1}{2}0.0964=0.0859$</code></li>
<li>For S: <code>$0.4247+\frac{1}{2}0.3343+\frac{1}{2}0.0964=0.64005$</code></li>
<li>For F: <code>$0.0843+\frac{1}{2}0.3343+\frac{1}{2}0.0452=0.27405$</code></li>
</ul>
<p>Checking, these three do sum to 1: <code>$0.0859+0.64005+0.27405=1$</code>.</p>
<h3 id="Problem_15"><a class="hanchor" href="#Problem_15">Problem 1.5</a></h3>
<pre><code>setrgb(0;0;0)
grid([0.00001 1 0.1];[0 10 1])
xtitle("q")
ytitle("Ratio of heterozygotes to homozygotes")

plot({(2*(1-x))%x})

setrgb(0;0;1)
plot({(2*(1-x))%x})
text(200;200;"2p/q")

setrgb(1;0;0)
plot({2%x})
text(250;250;"2/q")

draw()
</code></pre>
<p><img alt="Graph for 1.5" src="./img/pg_solutions/p1.5.png"/></p>
<h3 id="Problem_16"><a class="hanchor" href="#Problem_16">Problem 1.6</a></h3>
<pre><code>&gt;&gt;&gt; import numpy as np
&gt;&gt;&gt; import scipy.stats as sps
&gt;&gt;&gt; freq=np.array([0.4247, 0.3343, 0.0843, 0.0964, 0.0452, 0.0151])
&gt;&gt;&gt; expected=np.array([0.4096, 0.3507, 0.0751, 0.1101, 0.0471, 0.0074])
&gt;&gt;&gt; sps.chisquare(freq, expected)
Power_divergenceResult(statistic=0.012244149537675565, pvalue=0.9999991214396423)
</code></pre>
<p>I guess? But checking the solution, I‚Äôm apparently wrong in multiple ways:
3 degrees of freedom, and the genotype frequencies need to be multiplied
by 332 before conducting the test.</p>
<p>So, try again:</p>
<pre><code>&gt;&gt;&gt; import numpy as np
&gt;&gt;&gt; import scipy.stats as sps
&gt;&gt;&gt; freq=332*np.array([0.4247, 0.3343, 0.0843, 0.0964, 0.0452, 0.0151])
&gt;&gt;&gt; expected=332*np.array([0.4096, 0.3507, 0.0751, 0.1101, 0.0471, 0.0074])
&gt;&gt;&gt; sps.chisquare(freq, expected, ddof=3)
Power_divergenceResult(statistic=4.065057646508287, pvalue=0.13100381641660697)
</code></pre>
<p>Which is what the solution to the exercise says as well. I guess the
problem is that I don‚Äôt really know how to apply statistical tests
(‚Ä¶yet. Growth mindset.)</p>
<h2 id="Chapter_2"><a class="hanchor" href="#Chapter_2">Chapter 2</a></h2>
<h3 id="Problem_22"><a class="hanchor" href="#Problem_22">Problem 2.2</a></h3>
<p>Written in Lua, and generalizing to more than 2 alleles.</p>
<pre><code>-- number of individuals
N=260
-- number of generations
n=20
-- initial frequencies of the alleles
p={0.2, 0.4, 0.1, 0.1, 0.15, 0.05}
alleles={}
for i=1, #p do
    for j=1, math.floor(N*p[i]) do
        alleles[#alleles+1]="a"..i
    end
end
for i=1, n do
    nalleles={}
    for j=1, N/2 do
        a=alleles[math.random(#alleles)]
        nalleles[#nalleles+1]=a
        nalleles[#nalleles+2]=a
    end
    alleles=nalleles
end
table.sort(alleles)
print(table.concat(alleles, ", "))
</code></pre>
<h3 id="Problem_25"><a class="hanchor" href="#Problem_25">Problem 2.5</a></h3>
<pre><code>p2.5::.oc("p2.5.eps")
.tc(p2.5)

setrgb(0;0;0)
grid([1 100 10];[0 150 30])

xtitle("N")
ytitle("t¬Ω")

setrgb(0;0;1)
plot({-ln(2)%ln(1-1%2*x)})
text(200;160;"-ln(2)/ln(1-1/2N)")

setrgb(1;0;0)
plot({2*x*ln(2)})
text(250;275;"2Nln(2)")

draw()

.fl()
.cc(p2.5)
</code></pre>
<p><img alt="Graph for 2.5" src="./img/pg_solutions/p2.5.png"/></p>
<p>These are, in fact, two different graphs, but the approximation is good
enough that the difference between the two is not visible.</p>
<h3 id="Problem_27"><a class="hanchor" href="#Problem_27">Problem 2.7</a></h3>
<p>That would be <code>$3*3000$</code> for one step (3 possible changes for every
nucleotides).</p>
<p>For two mutational steps, it would be <code>$3^2*{3000 \choose 2}$</code> (3 possible
changes per chosen nucleotide, and two different nucleotides chosen from
the whole allele).</p>
<p>For <code>$n$</code> mutational steps, it would be <code>$3^n*{3000 \choose n}$</code>.</p>
<h3 id="Problem_29"><a class="hanchor" href="#Problem_29">Problem 2.9</a></h3>
<p>(The actual code doesn‚Äôt contain the unicode symbols since either Klong
or Postscript can‚Äôt deal with them. Sad.)</p>
<pre><code>p2.9::.oc("p2.9.eps")
.tc(p2.9)

setrgb(0;0;0)
grid([0 1 0.1];[0 0.0001 0.00001])

xtitle("ùìó")
ytitle("-Œî_Nùìó and Œî·µ§ùìó")

N::10^4
u::5*10^-5

setrgb(0;0;1)
plot({(1%2*N)*x})
text(200;200;"-Œî_Nùìó")

setrgb(1;0;0)
plot({2*u*1-x})
text(250;250;"Œî·µ§ùìó")

draw()

.fl()
.cc(p2.9)
</code></pre>
<p><img alt="Graph for 2.9" src="./img/pg_solutions/p2.9.png"/></p>
<p>Does the same thing fall out of the math?</p>
<p>Let‚Äôs see:</p>
<p>At least the intersection does.</p>
</body></html>
