
<title>niplav</title>
<link rel="shortcut icon" type="image/png" href="./favicon.png">
<link rel="stylesheet" type="text/css" href="main.css">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<!DOCTYPE HTML>

<style TYPE="text/css">
code.has-jax {font: inherit; font-size: 100%; background: inherit; border: inherit;}
</style>

<script type="text/javascript" async
	src="./mathjax/latest.js?config=TeX-MML-AM_CHTML">
</script>

<script type="text/x-mathjax-config">
	MathJax.Hub.Config({
	extensions: ["tex2jax.js"],
	jax: ["input/TeX", "output/HTML-CSS"],
	tex2jax: {
		inlineMath: [ ['$','$'], ["\\(","\\)"] ],
		displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
		processEscapes: true,
		skipTags: ['script', 'noscript', 'style', 'textarea', 'pre']
	},
	"HTML-CSS": { availableFonts: ["TeX"] }
	});
</script>

<script>
var anchorhash={}
function addAnchor(element) {
	var cnt=element.textContent
	if(cnt==="home")
		return;
	var ref=element.textContent.replace(/[^a-zA-Z0-9 ]/mg, "")
	ref=ref.replace(/ /mg, "-")
	var newref=ref;
	if(anchorhash[ref]===1)
		for(i=1, newref=ref+"_"+i;anchorhash[ref+"_"+i]===1;i++, newref=ref+"_"+i)
			;
	ref=newref
	element.setAttribute("id", `${ref}`)
	element.innerHTML=`<a href="#${ref}" class="hanchor">${cnt}</a>`
	anchorhash[ref]=1
}
document.addEventListener('DOMContentLoaded', function () {
	// Add anchor links to all headings
	var headers = document.querySelectorAll('h1, h2, h3, h4, h5, h6')
	if (headers) {
		headers.forEach(addAnchor)
	}
	// Change the title to the h1 header
	var title = document.querySelector('h1')
	if(title) {
		var title_elem = document.querySelector('title')
		title_elem.textContent=title.textContent + " ‚Äì niplav"
	}
});
</script>

<h2><a href="./index.html">home</a></h2>

<p><em>author: niplav, created: 2022-01-25, modified: 2022-02-23, language: english, status: in progress, importance: 2, confidence: likely</em></p>

<blockquote>
  <p><strong>Some solutions to exercises in the book ‚ÄúPopulation Genetics‚Äù by
<a href="https://en.wikipedia.org/wiki/John_H._Gillespie">John H. Gillespie</a>. I
did not simply copy out the solutions at the end of each chapter, and
sometimes didn't even check the solutions against my own. Therefore,
these might be faulty.</strong></p>
</blockquote>

<h1>Solutions to ‚ÄúPopulation Genetics‚Äù</h1>

<h2>Chapter 1</h2>

<h3>Problem 1.1</h3>

<p>For three it would be <code>$A_1A_1A_1, A_1A_1A_2, A_1A_2A_2, A_1A_2A_3, A_2A_2A_2, A_2A_2A_3, A_3A_3A_3$</code>,
with 7 different genotypes.</p>

<p>For four it would be 32 different genotypes:</p>

<p><code>$A_1 A_1 A_1 A_1, A_1 A_1 A_1 A_2, A_1 A_1 A_1 A_3, A_1 A_1 A_1 A_4, A_1 A_1 A_2 A_2, A_1 A_1 A_2 A_3, A_1 A_1 A_2 A_4, A_1 A_1 A_3 A_3, A_1 A_1 A_3 A_4, A_1 A_1 A_4 A_4, A_1 A_2 A_2 A_2, A_1 A_2 A_2 A_3, A_1 A_2 A_2 A_4, A_1 A_2 A_3 A_4, A_1 A_2 A_4 A_4, A_1 A_3 A_4 A_4, A_1 A_4 A_4 A_4, A_2 A_2 A_2 A_2, A_2 A_2 A_2 A_3, A_2 A_2 A_2 A_4, A_2 A_2 A_3 A_3, A_2 A_2 A_3 A_4, A_2 A_2 A_4 A_4, A_2 A_3 A_3 A_3, A_2 A_3 A_3 A_4, A_2 A_3 A_4 A_4, A_2 A_4 A_4 A_4, A_3 A_3 A_3 A_3, A_3 A_3 A_3 A_4, A_3 A_3 A_4 A_4, A_3 A_4 A_4 A_4, A_4 A_4 A_4 A_4$</code></p>

<!--TODO: what closed form formula describes the above counting
partition?-->

<p>So apparently, per solution, this is wrong, because I assumed that <code>$n$</code>
different alleles resulted in a <code>$n$</code>ploid organism, which isn't the case.</p>

<p>For three alleles it would be <code>$A_1 A_1, A_1 A_2, A_1 A_3, A_2 A_2, A_2 A_3, A_3 A_3$</code>,
which is 6 different genotypes.</p>

<p>For four different alleles it would be 10 different genotypes:</p>

<p><code>$A_1 A_1,A_1 A_2,A_1 A_3,A_1 A_4,A_2 A_2,A_2 A_3,A_2 A_4,A_3 A_3,A_3 A_4,A_4 A_4$</code></p>

<p>The general formula is <code>$\frac{n (n+1)}{2}$</code> different genotypes,
because it's just the upper half of the square again.</p>

<h3>Problem 1.2</h3>

<ul>
<li>For I: <code>$0.0151+\frac{1}{2}0.0452+\frac{1}{2}0.0964=0.0859$</code></li>
<li>For S: <code>$0.4247+\frac{1}{2}0.3343+\frac{1}{2}0.0964=0.64005$</code></li>
<li>For F: <code>$0.0843+\frac{1}{2}0.3343+\frac{1}{2}0.0452=0.27405$</code></li>
</ul>

<p>Checking, these three do sum to 1: <code>$0.0859+0.64005+0.27405=1$</code>.</p>

<h3>Problem 1.5</h3>

<pre><code>setrgb(0;0;0)
grid([0.00001 1 0.1];[0 10 1])
xtitle("q")
ytitle("Ratio of heterozygotes to homozygotes")

plot({(2*(1-x))%x})

setrgb(0;0;1)
plot({(2*(1-x))%x})
text(200;200;"2p/q")

setrgb(1;0;0)
plot({2%x})
text(250;250;"2/q")

draw()
</code></pre>

<p><img src="./img/pg_solutions/p1.5.png" alt="Graph for 1.5" title="" /></p>

<h2>Chapter 2</h2>

<h3>Problem 2.2</h3>

<p>Written in Lua, and generalizing to more than 2 alleles.</p>

<pre><code>-- number of individuals
N=260
-- number of generations
n=20
-- initial frequencies of the alleles
p={0.2, 0.4, 0.1, 0.1, 0.15, 0.05}
alleles={}
for i=1, #p do
    for j=1, math.floor(N*p[i]) do
        alleles[#alleles+1]="a"..i
    end
end
for i=1, n do
    nalleles={}
    for j=1, N/2 do
        a=alleles[math.random(#alleles)]
        nalleles[#nalleles+1]=a
        nalleles[#nalleles+2]=a
    end
    alleles=nalleles
end
table.sort(alleles)
print(table.concat(alleles, ", "))
</code></pre>

<h3>Problem 2.5</h3>

<pre><code>p2.5::.oc("p2.5.eps")
.tc(p2.5)

setrgb(0;0;0)
grid([1 100 10];[0 150 30])

xtitle("N")
ytitle("t¬Ω")

setrgb(0;0;1)
plot({-ln(2)%ln(1-1%2*x)})
text(200;160;"-ln(2)/ln(1-1/2N)")

setrgb(1;0;0)
plot({2*x*ln(2)})
text(250;275;"2Nln(2)")

draw()

.fl()
.cc(p2.5)
</code></pre>

<p><img src="./img/pg_solutions/p2.5.png" alt="Graph for 2.5" title="" /></p>

<p>These are, in fact, two different graphs, but the approximation is good
enough that it's not visible.</p>

<h3>Problem 2.7</h3>

<p>That would be <code>$3*3000$</code> for one step (3 possible changes for every
nucleotides).</p>

<p>For two mutational steps, it would be <code>$3^2*{3000 \choose 2}$</code> (3 possible
changes per chosen nucleotide, and two different nucleotides chosen from
the whole allele).</p>

<p>For <code>$n$</code> mutational steps, it would be <code>$3^n*{3000 \choose n}$</code>.</p>

<h3>Problem 2.9</h3>

<p>(The actual code doesn't contain the unicode symbols since either Klong
or Postscript can't deal with them. Sad.)</p>

<pre><code>p2.9::.oc("p2.9.eps")
.tc(p2.9)

setrgb(0;0;0)
grid([0 1 0.1];[0 0.0001 0.00001])

xtitle("ùìó")
ytitle("-Œî_Nùìó and Œî·µ§ùìó")

N::10^4
u::5*10^-5

setrgb(0;0;1)
plot({(1%2*N)*x})
text(200;200;"-Œî_Nùìó")

setrgb(1;0;0)
plot({2*u*1-x})
text(250;250;"Œî·µ§ùìó")

draw()

.fl()
.cc(p2.9)
</code></pre>

<p><img src="./img/pg_solutions/p2.9.png" alt="Graph for 2.9" title="" /></p>

<p>Does the same thing fall out of the math?</p>

<p>Let's see:</p>

<div>
    $$-Œî_N \mathcal{H}=Œî_u \mathcal{H} \Leftrightarrow \\
    \frac{1}{2N} \mathcal{H}=2u(1-\mathcal{H}) \Leftrightarrow \\
    \mathcal{H}=4uN-4uN \mathcal{H} \Leftrightarrow \\
    \mathcal{H}+4uN \mathcal{H}=4uN \Leftrightarrow \\
    \mathcal{H}=\frac{4uN}{1+4uN} \Leftrightarrow \\
    \mathcal{H}=0.\overline{6} $$
</div>

<p>At least the intersection does.</p>
</html>
